File: Guigna-Swift/Guigna/Systems/MacOSX.swift
import Foundation

class MacOSX: GSystem {
    
    init(agent: GAgent) {
        super.init(name: "Mac OS X", agent: agent)
        prefix = "" // TODO
        homepage = "http://support.apple.com/downloads/"
        cmd = "/usr/sbin/pkgutil"
    }
    
    override func list() -> [GPackage] {
        
        index.removeAll(keepCapacity: true)
        items.removeAll(keepCapacity: true)
        
        items = installed()
        return items as [GPackage]
    }
    
    
    override func installed() -> [GPackage] {
        
        var pkgs = [GPackage]()
        pkgs.reserveCapacity(1000)
        
        var pkgIds = output("/usr/sbin/pkgutil --pkgs").split("\n")
        pkgIds.removeLast()
        
        let history = (NSArray(contentsOfFile: "/Library/Receipts/InstallHistory.plist") as Array).reverse()
        var keepPkg: Bool
        for dict in history as [NSDictionary] {
            keepPkg = false
            var ids = dict["packageIdentifiers"]! as [String]
            for pkgId in ids {
                if let idx = find(pkgIds, pkgId) {
                    keepPkg = true
                    pkgIds.removeAtIndex(idx)
                }
            }
            if !keepPkg {
                continue
            }
            let name = dict["displayName"]! as String
            var version = dict["displayVersion"]! as String
            var category = dict["processName"]! as String
            category = category.stringByReplacingOccurrencesOfString(" ", withString: "").lowercaseString
            if category == "installer" {
                let plist = output("/usr/sbin/pkgutil --pkg-info-plist \(ids[0])").propertyList() as NSDictionary
                version = plist["pkg-version"]! as String
            }
            var pkg = GPackage(name: name, version: "", system: self, status: .UpToDate)
            pkg.id = ids.join()
            pkg.categories = category
            pkg.description = pkg.id!
            pkg.installed = version
            // TODO: pkg.version
            pkgs += pkg
        }
        //    for pkg in installed() {
        //        index[pkg key].status = pkg.status
        //    }
        return pkgs
    }
    
    
    override func outdated() -> [GPackage] {
        var pkgs = [GPackage]()
        // TODO: sudo /usr/sbin/softwareupdate --list
        return pkgs
    }
    
    
    override func inactive() -> [GPackage] {
        var pkgs = [GPackage]()
        return pkgs
    }
    
    
    override func info(item: GItem) -> String {
        var info = ""
        for pkgId in item.id.split() {
            info += output("/usr/sbin/pkgutil --pkg-info \(pkgId)")
            info += "\n"
        }
        return info
    }
    
    override func home(item: GItem) -> String {
        var homeppage = "http://support.apple.com/downloads/"
        if item.categories == "storeagent" || item.categories == "storedownloadd" {
            let url = "http://itunes.apple.com/lookup?bundleId=\(item.id)"
            let data = NSData(contentsOfURL: NSURL(string: url))
            let results = ((NSJSONSerialization.JSONObjectWithData(data, options: nil, error: nil) as NSDictionary)["results"]! as NSArray)
            if results.count > 0 {
                let pkgId = (results[0] as NSDictionary)["trackId"]!.stringValue!
                let url = NSURL(string: "http://itunes.apple.com/app/id\(pkgId)")
                let xmlDoc = NSXMLDocument(contentsOfURL: url, options: Int(NSXMLDocumentTidyHTML), error: nil)
                let mainDiv = xmlDoc.rootElement().nodesForXPath("//div[@id=\"main\"]", error: nil)[0] as NSXMLNode
                let links = mainDiv["//div[@class=\"app-links\"]/a"]
                // TODO: get screenshots via JSON
                let screenshotsImgs = mainDiv["//div[contains(@class, \"screenshots\")]//img"]
                var screenshots = ""
                var i: Int = 0
                for img in screenshotsImgs {
                    let url = img.attribute("src")
                    if i > 0 {
                        screenshots += " "
                    }
                    screenshots += url
                    i++
                }
                item.screenshots = screenshots
                homepage = links[0].attribute("href")
                if homepage == "http://" {
                    homepage = links[1].attribute("href")
                }
            }
        }
        return homepage
    }
    
    override func log(item: GItem!) -> String {
        var page = "http://support.apple.com/downloads/"
        if item != nil {
            if item.categories == "storeagent" || item.categories == "storedownloadd" {
                let url = "http://itunes.apple.com/lookup?bundleId=\(item.id)"
                let data = NSData(contentsOfURL: NSURL(string: url))
                let results = ((NSJSONSerialization.JSONObjectWithData(data, options: nil, error: nil) as NSDictionary)["results"]! as NSArray)
                if results.count > 0 {
                    let pkgId = (results[0] as NSDictionary)["trackId"]!.stringValue!
                    page = "http://itunes.apple.com/app/id" + pkgId
                }
            }
        }
        return page
    }
    
    override func contents(item: GItem) -> String {
        var contents = ""
        for pkgId in item.id.split() {
            let plist = output("\(cmd) --pkg-info-plist \(pkgId)").propertyList() as NSDictionary
            var files = output("\(cmd) --files \(pkgId)").split("\n")
            files.removeLast()
            for file in files {
                contents += NSString.pathWithComponents([plist["volume"]!, plist["install-location"]!, file])
                contents += ("\n")
            }
        }
        return contents
    }
    
    override func cat(item: GItem) -> String {
        return "TODO"
    }
    
    
    override func uninstallCmd(pkg: GPackage) -> String {
        // SEE: https://github.com/caskroom/homebrew-cask/blob/master/lib/cask/pkg.rb
        var commands = [String]()
        let fileManager = NSFileManager.defaultManager()
        var dirsToDelete = [String]()
        var isDir: ObjCBool = true
        for pkgId in pkg.id.split() {
            let plist = output("\(cmd) --pkg-info-plist \(pkgId)").propertyList() as NSDictionary
            var dirs = output("\(cmd) --only-dirs --files \(pkgId)").split("\n")
            dirs.removeLast()
            for dir in dirs {
                let dirPath: NSString = NSString.pathWithComponents([plist["volume"]!, plist["install-location"]!, dir])
                let fileAttributes = fileManager.attributesOfItemAtPath(dirPath, error: nil) as NSDictionary
                if (!(fileAttributes.fileOwnerAccountID() == 0) && !dirPath.hasPrefix("/usr/local"))
                    || dirPath.containsString(pkg.name)
                    || dirPath.containsString(".")
                    || dirPath.hasPrefix("/opt/") {
                        if (dirsToDelete.filter { dirPath.containsString($0) }).count == 0 {
                            dirsToDelete += dirPath
                            commands += "sudo rm -r \"\(dirPath)\""
                        }
                }
            }
            var files = output("\(cmd) --files \(pkgId)").split("\n") // links are not detected with --only-files
            files.removeLast()
            for file in files {
                let filePath = NSString.pathWithComponents([plist["volume"]!, plist["install-location"]!, file])
                if !(fileManager.fileExistsAtPath(filePath, isDirectory: &isDir) && (isDir == true)) {
                    if (dirsToDelete.filter { filePath.contains($0) }).count == 0 {
                        commands += "sudo rm \"\(filePath)\""
                    }
                }
            }
            commands += "sudo \(cmd) --forget \(pkgId)"
        }
        return commands.join(" ; ")
        
        // TODO: disable Launchd daemons, clean Application Support, Caches, Preferences
        // SEE: https://github.com/caskroom/homebrew-cask/blob/master/lib/cask/artifact/pkg.rb
    }
    
}

File: Guigna-RubyMotion/app/lib/systems/macosx.rb
# require 'GAdditions'
# require 'GuignaItems'
# require 'GAgent'
# require 'GuignaSystems'

class MacOSX < GSystem # TODO: review
  
  @prefix = "/usr/sbin"
  
  def initialize(agent=nil)
    super("Mac OS X", agent)
    @homepage = "http://support.apple.com/downloads/"
    @cmd = "#{@prefix}/pkgutil"
  end
  def list
    @index.clear
    @items.clear
    @items = self.installed
  end
  def installed
    pkgs = []
    pkgs_ids = `/usr/sbin/pkgutil --pkgs`.split("\n")
    history = load_plist(File.read("/Library/Receipts/InstallHistory.plist")).reverse!
    for dict in history
      keep_pkg = false
      ids = dict['packageIdentifiers']
      for pkg_id in ids
        if pkgs_ids.include?(pkg_id)
          keep_pkg = true
          pkgs_ids.delete pkg_id
        end
      end
      next if !keep_pkg
      name = dict['displayName']
      version = dict['displayVersion']
      category = dict['processName'].gsub(" ","").downcase
      if category == "installer"
        version = load_plist(`#{cmd} --pkg-info-plist #{ids[0]}`)["pkg-version"]
      end
      pkg = GPackage.new(name, nil, self, :uptodate)
      pkg.id = ids.join(' ')
      pkg.categories = category
      pkg.description = pkg.id
      pkg.installed = version # TODO: pkg.version
      pkgs << pkg
      # @index[pkg.key] = pkg
      # TODO verify installed
    end
    pkgs
  end
  
  # TODO
  def outdated
    pkgs = []
    return pkgs
  end
  
  def info(pkg)
    output = ""
    pkg.id.split.each do |id|
      output << `#{cmd} --pkg-info #{id}` + "\n"
    end
    output
  end
  
  def home(pkg)
    homepage = "http://support.apple.com/downloads/"
    if pkg.categories == "storeagent" || pkg.categories == “storedownloadd”
      url = "http://itunes.apple.com/lookup?bundleId=#{pkg.id}"
      data = NSData.dataWithContentsOfURL(NSURL.URLWithString url)
      results = NSJSONSerialization.JSONObjectWithData(data, options:0, error:nil)["results"]
      if results.size > 0
        pkg_id = results[0]["trackId"].to_s
        main_div = agent.nodes_for_url("http://itunes.apple.com/app/id" + pkg_id, xpath:"//div[@id=\"main\"]").first
        links = main_div["//div[@class=\"app-links\"]/a"]
        screenshots_imgs = main_div["//div[contains(@class, \"screenshots\")]//img"]
        pkg.screenshots = screenshots_imgs.map {|img| img['@src']}.join(" ")
        homepage = links[0].href
        homepage = links[1].href if homepage == "http://"
      end
    end
    return homepage
  end
  
  def log(pkg)
    page = "http://support.apple.com/downloads/"
    if !pkg.nil?
      if pkg.categories == "storeagent" || pkg.categories == “storedownloadd”
        url = "http://itunes.apple.com/lookup?bundleId=#{pkg.id}"
        data = NSData.dataWithContentsOfURL(NSURL.URLWithString url)
        results = NSJSONSerialization.JSONObjectWithData(data, options:0, error:nil)["results"]
        if results.size > 0
          pkg_id = results[0]["trackId"].to_s
          page = "http://itunes.apple.com/app/id" + pkg_id
        end
      end
    end
    return page
  end
  
  def contents(pkg)
    contents = ""
    pkg.id.split.each do |id|
      plist = load_plist `#{cmd} --pkg-info-plist #{id}`
      files = `#{@cmd} --files #{id}`.split("\n")
      files.each do |file|
        contents << NSString.pathWithComponents([plist['volume'], plist['install-location'], file]) +"\n"
      end
    end
    contents
  end
  
  
  def uninstall_cmd(pkg)
    # SEE: https://github.com/caskroom/homebrew-cask/blob/master/lib/cask/pkg.rb
    commands = []
    dirs_to_delete = []
    for pkg_id in pkg.id.split
      plist = load_plist `#{cmd} --pkg-info-plist #{pkg_id}`
      dirs = `#{cmd} --only-dirs --files #{pkg_id}`.split("\n")
      for dir in dirs
        dir_path = NSString.pathWithComponents([plist['volume'], plist['install-location'], dir])
        
        if (File.stat(dir_path).uid != 0 && !dir_path.start_with?('/usr/local')) \
          || dir_path.include?(pkg.name) || dir_path.include?('.') \
          || dir_path.start_with?('/opt/')
          
          if dirs_to_delete.select {|d| dir_path.include?(d)}.size == 0
            dirs_to_delete << dir_path
            commands << "sudo rm -r \"#{dir_path}\""
          end
        end
      end
      files = `#{cmd} --files #{pkg_id}`.split("\n") # links are not detected with --only-files
      for file in files
        file_path = NSString.pathWithComponents([plist['volume'], plist['install-location'], file])
        unless File.directory?(file_path)
          if dirs_to_delete.select {|d| file_path.include?(d)}.size == 0
            commands << "sudo rm \"#{file_path}\""
          end
        end
      end
      commands << "sudo #{cmd} --forget #{pkg_id}"
    end
    return commands.join(" ; ")
  end
  # TODO: disable Launchd daemons, clean Application Support, Caches, Preferences
  # SEE: https://github.com/caskroom/homebrew-cask/blob/master/lib/cask/artifact/pkg.rb
end
